doc/dev/libraries.md

```diff
@@ -0,0 +1,84 @@
+- `libelektra-core`:
+  The core minimal API of Elektra.
+  Defines what `ElektraKey` and `ElektraKeyset` are, and contains the minimal API for manipulating them.
+
+  **Prefixes:** `elektraKey*` and `elektraKeyset*`
+
+- `libelektra-kdb`:
+  The main API for interfacing with the KDB.
+
+  **Prefix:** `elektraKdb*`
+
+- `libelektra-lowlevel-c`:
+  Additional C APIs that are useful when working with `ElektraKey` and `ElektraKeyset`, but not are not minimal API.
+  Specifically targets C and not intended for use via bindings.
+
+  **Prefix:** `elektra*`
+
+- `libelektra-highlevel-c`:
+  The C high-level API for reading/writing configuration with Elektra.
+  Intended for use by applications.
+  Partially intended for use with code-generation.
+  Specifically targets C and not intended for use via bindings.
+
+  **Prefix:** `elektraHlc*` (better suggestions welcome)
+  (Note: Even though there isn't any other `highlevel` implementation, I added the `-c` suffix to show this lib is only for C)
+
+- `libelektra-opts`:
+  The API for parsing command-line arguments according to Elektra's spec.
+  **Prefix:** `elektraOpts*`
+- `libelektra-notification`:
+  The API for setting up callbacks and automatically updated variables linked to changes in the KDB.
+
+  **Prefix:** `elektraNotification*`
+
+- `libelektra-io`:
+  Elektra's API for asynchronous operations.
+
+  **Prefix:** `elektraIo*`
+
+- `libelektra-merge`:
+  The API for merging and/or detecting conflicts between two keysets.
+
+  **Prefix:** `elektraMerge`
+
+- `libelektra-pluginbase`:
+  The API _used by_ plugins.
+
+  **Prefix:** `elektraPlugin*`
+  (Note: previously `libelektra-plugin`)
+
+- `libelektra-plugin`:
+  The API for _interacting with_ plugins.
+  This includes loading plugins and calling exported functions.
+
+  **Prefix:**: `elektraPlugin*`
+  (Note: previously `libelektra-invoke` & `loader`)
+
+- `libelektra-mount`:
+  The API for manipulating mountpoints.
+
+  **Prefix:** `elektraMount*`
+
+- `libelektra-type`:
+  The API that defines Elektra's type system. (Alternative: The API that defines the type system used by `libelektra-highlevel-c`.)
+
+  **Names:** `elektra<TYPE>ToString` and `elektraKeyTo<TYPE>`
+  (Note: extracted from `libelektra-ease`)
+
+- `libelektra-ease`:
+  A collection of various other APIs that help when interacting with `ElektraKey` and `ElektraKeyset`.
+  Contains APIs that are not large, specific, or complex enough to justify a separate library and not minimal enough for `libelektra-core`, but can be implemented on top of the public API.
+  The APIs should not specifically target C and should be usable via bindings (if appropriate for the other language).
+
+  **Prefix:** `elektra*`
```

Comment:
I would prefer if we can be consistent and always have (several) specific prefixes for all libs.

Reply:
Maybe as a compromise: 

- `libelektra-ease` **Prefix:** `elektraEase*`
- `libelektra-operations` **Prefix:** `elektraOps*`, or `elektraKey*`/`elektraKeyset*` when it doesn't conflict with `libelektra-core`
- `libelektra-lowlevel-c` **Prefix:** `elektraLlc*`, or `elektraKey*`/`elektraKeyset*` when it doesn't conflict with `libelektra-core`

The reason why I would allow `elektraKey*`/`elektraKeyset*` in `libelektra-operations` and `libelektra-lowlevel-c` are things like `elektraKeysetCut`. Calling it `elektraOpsKeysetCut` would hurt discoverability and development experience. It is much nicer IMO, if everything that does some sort of concrete operation on a `Key`/`Keyset` has the same prefix.

There shouldn't ever be a name conflict, because if a function is minimal it should be in the core, otherwise it is not. And between operations and lowlevel-c the conflict is resolved by looking at whether the function is useful outside the C world.

Reply:
> If we link `libelektra-operations` by default

We can't really define what gets linked that's on the user side. But yes, I would include `libelektra-operations` (and maybe `libelektra-lowlevel-c`) in the pkgconfig and CMake configs that we provide.

> I would rather make exceptions for the high-level API, as this API is much more user-facing then these low level libraries.

Yes, that's why I didn't really like the `elektraHlc*` prefix. But maybe we can use `elektraConf*` for the high-level API. That's still a pretty good prefix for use in applications. We definitely need a prefix beyond `elektra*`, because otherwise it becomes hard to see what is part of the high-level API (i.e. intended for applications) and what is more low-level.

Reply:
>  We can't really define what gets linked that's on the user side. But yes, I would include libelektra-operations (and maybe libelektra-lowlevel-c) in the pkgconfig and CMake configs that we provide.

Yes, I mean the pkgconfig and CMake configs to be "default". Is there anyone not using them?

> elektraConf* 

Yes, that would be a possibility.

Reply:
> Is there anyone not using them?

Who knows what kinds of proprietary internal build systems are out there.

To come back to the original issue: We want to keep `elektraKey*` and `elektraKeyset*` exclusive to `libelektra-core` while having a nice convention for `libelektra-operations`. How about using `elektraXKey*` and `elektraXKeyset*` (`X` for extension/extra)?

For potential other functions in `libelektra-operations` that are not extensions of the Key or Keyset APIs, I'd simply propose `elektraOps*`.